src/irc/core/sasl.c

```diff
@@ -81,76 +162,131 @@ static void sasl_fail(IRC_SERVER_REC *server, const char *data, const char *from
 }
 
 static void sasl_success(IRC_SERVER_REC *server, const char *data, const char *from)
 {
-	if (server->sasl_timeout != 0) {
-		g_source_remove(server->sasl_timeout);
-		server->sasl_timeout = 0;
-	}
+	sasl_timeout_cancel(server);
 
 	signal_emit("server sasl success", 1, server);
 
 	/* The authentication succeeded, time to finish the CAP negotiation */
 	cap_finish_negotiation(server);
 }
 
-static void sasl_step(IRC_SERVER_REC *server, const char *data, const char *from)
+static void sasl_send_step_if_complete(IRC_SERVER_REC *server, const char *enc_req)
 {
-	IRC_SERVER_CONNECT_REC *conn;
-	GString *req;
-	char *enc_req;
-
-	conn = server->connrec;
-
-	/* Stop the timer */
-	if (server->sasl_timeout != 0) {
-		g_source_remove(server->sasl_timeout);
-		server->sasl_timeout = 0;
+	char *buffer;
+	size_t enc_req_len;
+
+	enc_req_len = strlen(enc_req);
+
+	buffer = g_hash_table_lookup(sasl_buffers, server->tag);
+	if (buffer != NULL) {
+		if (!g_strcmp0("+", enc_req)) {
+			enc_req = buffer;
+		} else {
+			enc_req = g_strconcat(buffer, enc_req, NULL);
+		}
+		g_hash_table_remove(sasl_buffers, server->tag);
 	}
 
-	switch (conn->sasl_mechanism) {
-		case SASL_MECHANISM_PLAIN:
-			/* At this point we assume that conn->sasl_{username, password} are non-NULL.
-			 * The PLAIN mechanism expects a NULL-separated string composed by the authorization identity, the
-			 * authentication identity and the password.
-			 * The authorization identity field is explicitly set to the user provided username.
-			 * The whole request is then encoded in base64. */
-
-			req = g_string_new(NULL);
+	if (enc_req_len == 400) {
+		g_hash_table_insert(sasl_buffers, server->tag, g_strdup(enc_req));
+	} else {
+		gchar *output;
+		GBytes *decoded;
+		IRC_SERVER_CONNECT_REC *conn;
+
+		if (!g_strcmp0("+", enc_req)) {
+			decoded = g_bytes_new("", 0);
+		} else {
+			gsize dec_len;
+			guchar *tmp;
+
+			tmp = g_base64_decode(enc_req, &dec_len);
+			decoded = g_bytes_new(tmp, dec_len);
+		}
+
+		conn = server->connrec;
+		output = g_strconcat("server sasl step ", conn->sasl_mechanism, NULL);
+		ascii_strdown(output+17);
+
+		if (!signal_emit(output, 2, server, decoded)) {
+			g_warning("No one can handle SASL mechanism: %s",
+				  conn->sasl_mechanism);
+		}
+		g_free(output);
+		g_bytes_unref(decoded);
+	}
 
-			g_string_append(req, conn->sasl_username);
-			g_string_append_c(req, '\0');
-			g_string_append(req, conn->sasl_username);
-			g_string_append_c(req, '\0');
-			g_string_append(req, conn->sasl_password);
+	g_free_not_null(buffer);
+}
 
-			enc_req = g_base64_encode((const guchar *)req->str, req->len);
+void sasl_send_response(IRC_SERVER_REC *server, GBytes *response)
+{
+	char *enc;
+	const guchar *p;
+	gsize len;
+	size_t offset, enc_len;
+
+	p = g_bytes_get_data(response, &len);
+	enc = g_base64_encode(p, len);
+	enc_len = strlen(enc);
+
+	offset = 0;
+	while (offset < enc_len) {
+		irc_send_cmdv(server, "AUTHENTICATE %400s",
```

Comment:
It's probably better to format the payload here instead of relying on the printf.